src/zod.ts

```diff
@@ -0,0 +1,211 @@
+import { z } from "zod";
+import { Action, LinkType, MatchMode } from "./constants.js";
+import { existsSync } from "fs";
+import { logger } from "./logger.js";
+import { getFileConfig } from "./configuration.js";
+
+const fileConfig = await getFileConfig();
+
+/**
+ * VALIDATION_SCHEMA is just a object of all the zod schemas
+ * each are named after what they are intended to validate
+ */
+
+export const VALIDATION_SCHEMA = z
+	.object({
+		notificationWebhookUrl: z.union([
+			z.undefined(),
+			z
+				.string()
+				.url()
+				.refine((url) => ({
+					message: `invalid notificationWebhookUrl URL - ${url}`,
+				})),
+		]),
+		torznab: z.array(
+			z
+				.string()
+				.url()
+				.refine((url) => ({
+					message: `invalid torznab URL - ${url}`,
+				}))
+		),
+		dataDirs: z.union([
+			z.undefined(),
+			z.array(
+				z.string().refine(
+					(path) => existsSync(path),
+					(path) => ({
+						message: `the dataDirs path ${path} does not exist on the filesystem`,
+					})
+				)
+			),
+		]),
+		torrentDir: z.string().refine(
+			(path) => existsSync(path),
```

Comment:
I think it might still behoove us to keep these dependencies on the filesystem separate from the schema validation—it feels like surprising behavior that calling a function named `parse` is going to spray `existsSync` calls. If we ever add unit tests down the road that would be harder to test well.

That said, I do see the benefit here of having a place we can check all provided directories to make sure they all resolve, and I think it's still worth doing that comprehensive check on all directories—just maybe after the schema validation stuff. What do you think though?

Reply:
While I understand the point of restricting the zod validation to the config file, I'm not sure this should really be concerning. We could always make another function that just calls parse, if the naming is the only issue.


Do you have any idea of where you would put the filesystem check instead? We could do it further down the line - perhaps just in doStartUpValidation after everything else is validated. checkConfigPaths() or something.

We should definitely do it early enough to terminate if they are invalid, though.

Reply:
yeah - I would do it further down the line in doStartUpValidation. checkConfigPaths makes sense to me 👍